File: src/modules/company/services/company_delete_service.py
from django.utils import timezone
from src.shared.types.user_request import UserRequest
from src.modules.company.models import Company
from src.shared.exceptions.not_found_exception import NotFoundException
from src.shared.validators.user_company_validator import (
    user_company_validator,
)
from src.modules.company.validators.company_delete_validator import (
    company_delete_validator,
)
from src.modules.company.services.company_delete_avatar_service import (
    CompanyDeleteAvatarService,
)


class CompanyDeleteService:
    @classmethod
    def execute(self, id: int, user_request: UserRequest):

        company = Company.objects.filter(id=id, deleted_at=None).first()

        if company is None:
            raise NotFoundException("Empresa não encontrada!")

        user_company_validator(id, user_request)
        company_delete_validator(id)

        if company.avatar is not None:
            company = CompanyDeleteAvatarService.execute(id, user_request)

        company.deleted_at = timezone.now()
        company.save()

        return company

File: src/modules/company/services/company-delete.service.ts
import { prisma } from '../../../infra/database/prisma.service';
import { NotFoundException } from '../../../shared/exceptions';
import { UserRequest } from '../../../shared/types/user-request';
import { userCompanyValidator } from '../../../shared/validators/user-company.validator';
import { CompanyEntity } from '../domain/entities/company.entity';
import { companyDeleteValidator } from '../validators/company-delete.validator';
import { companyDeleteAvatarService } from './company-delete-avatar.service';

class CompanyDeleteService {
  async execute(id: number, userRequest: UserRequest): Promise<CompanyEntity> {
    const company = await prisma.company.findFirst({
      where: {
        id: id,
        deletedAt: null,
      },
      include: {
        avatar: true,
      },
    });

    if (!company) throw new NotFoundException('Empresa não encontrada!');

    await userCompanyValidator(id, userRequest);

    await companyDeleteValidator(id);

    if (company.avatar)
      await companyDeleteAvatarService.execute(id, userRequest);

    const updatedCompany = await prisma.company.update({
      data: {
        deletedAt: new Date(),
        address: {
          update: {
            deletedAt: new Date(),
          },
        },
      },
      where: {
        id: id,
      },
    });

    return new CompanyEntity(updatedCompany);
  }
}

export const companyDeleteService = new CompanyDeleteService();
